testkit/src/mill/testkit/BspServerUtil.scala

```diff
@@ -0,0 +1,184 @@
+package mill.testkit
+
+import ch.epfl.scala.{bsp4j => b}
+import com.google.gson.{Gson, GsonBuilder}
+import coursier.cache.CacheDefaults
+import mill.api.BuildInfo
+import mill.bsp.Constants
+import org.eclipse.{lsp4j => l}
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{ExecutorService, Executors, ThreadFactory}
+
+import scala.jdk.CollectionConverters._
+import scala.reflect.ClassTag
+
+object BspServerUtil {
+
+  val updateFixtures = false
+
+  lazy val bsp4jVersion: String = sys.props.getOrElse("BSP4J_VERSION", ???)
+
+  trait DummyBuildClient extends b.BuildClient {
+    def onBuildLogMessage(params: b.LogMessageParams): Unit = ()
+    def onBuildPublishDiagnostics(params: b.PublishDiagnosticsParams): Unit = ()
+    def onBuildShowMessage(params: b.ShowMessageParams): Unit = ()
+    def onBuildTargetDidChange(params: b.DidChangeBuildTarget): Unit = ()
+    def onBuildTaskFinish(params: b.TaskFinishParams): Unit = ()
+    def onBuildTaskProgress(params: b.TaskProgressParams): Unit = ()
+    def onBuildTaskStart(params: b.TaskStartParams): Unit = ()
+    def onRunPrintStderr(params: b.PrintParams): Unit = ()
+    def onRunPrintStdout(params: b.PrintParams): Unit = ()
+  }
+
+  lazy val dummyBuildClient: b.BuildClient = new DummyBuildClient {}
```

Comment:
The convention we've used elsewhere in the codebase is `object DummyBuildClient extends DummyBuildClient`

Reply:
Sure (although bin compat might be harder to preserve with the `object`)